### PR TITLE
Create route for typeform submissions and post to Canvas submissions

### PR DESCRIPTION
**Pull request by apackin**
#### What's this PR do?
Creates a route for assignments submitted by typeform.
Add Canvas functionality for submitting assignments.
#### How should this be manually tested?
1: Create a test Canvas account with a Canvas assignment that accepts text based submission
2: node www/app.js 
3: post to localhost/canvas/typeformAssignment/:assignment
with a typeform like body (swap in the email associated with the canvas account):  `{
  "event_id": "UwVP9f9sGU",
  "event_type": "form_response",
  "form_response": {
    "form_id": "L5aBhG",
    "token": "875527530cbc8ec306774088db80887f",
    "submitted_at": "2016-08-12T14:49:14Z",
    "calculated": {
      "score": 15
    },
    "definition": {
      "id": "L5aBhG",
      "title": "",
      "fields": [
        {
          "id": "3",
          "title": "Question text: 3",
          "type": "textfield"
        },
        {
          "id": "2",
          "title": "Question Bool: 2>",
          "type": "yes-no"
        },
        {
          "id": "1",
          "title": "Please enter your email below",
          "type": "email"
        }
      ]
    },
    "answers": [
      {
        "type": "text",
        "text": "asdfg",
        "field": {
          "id": "3",
          "type": "short_text"
        }
      },
      {
        "type": "boolean",
        "boolean": true,
        "field": {
          "id": "2",
          "type": "yes_no"
        }
      },
      {
        "type": "email",
        "email": "abc@abc.com",
        "field": {
          "id": "1",
          "type": "email"
        }
      }
    ]
  }
}`

#### What are the relevant tickets?
https://admin.crisistextline.org/jira/secure/RapidBoard.jspa?rapidView=5&view=detail&selectedIssue=INT-223

### CODE REVIEW

**Comment by apackin** [comment]
Starting to write tests for this tomorrow, but would appreciate an eye on this before then.

**Comment by jrauschenberg** [comment]
It seems like I was able to hit the route, but I did get the following in the terminal, and the user's assignment has not, as yet, updated as submitted:
[Thu, Aug 18, 2016 12:40 PM -04:00] Error finding 1073 in course found for john.rauschenberg@gmail.com: Canvas returned 0 assignments
[Thu, Aug 18, 2016 12:40 PM -04:00] Canvas error message: [object Object] 200
POST /canvas/typeformAssignment/1073 200 737ms - 380b
[Thu, Aug 18, 2016 12:40 PM -04:00] [{"email":"heather@crisistextline.org","status":"sent","_id":"beb420a40012432a9bade8cb98446027","reject_reason":null}]

**Comment by apackin** [comment]
You need to use the assignment name in the URL instead of its ID

File: jobs/scheduling/helpers/updateCanvas.js [hunk 1]
```diff
@@ -105,24 +105,97 @@ canvas.updateUserGrade = function(user, course, assignment, grade) {
...
+        body: submissionHTML 
+      }
+    }
+  };
+
+  function callback(error, response){
+    if(!error && response.statusCode === 201){
+      CONSOLE_WITH_TIME(`Canvas submission succeeded for user ID ${user}`);
+    } else {
+      CONSOLE_WITH_TIME(`Canvas error message: ${response} ${response.statusCode}`);
```
Comment (jrauschenberg) [inline jobs/scheduling/helpers/updateCanvas.js]: Maybe add something here like `Canvas submission failed for user ID ${user}` before you go on to the error message?

File: jobs/scheduling/helpers/updateCanvas.js [hunk 2]
```diff
@@ -105,24 +105,97 @@ canvas.updateUserGrade = function(user, course, assignment, grade) {
...
+
+canvas.retrieveUserCourseAssignmentIds = function(userEmail, assignment, errFunc){
+  let promiseUserID;
+  let promiseCourseID;
+  let promiseAssignmentID;
+
+  // request canvas user with this email address and find their userId
+  promiseUserID = canvas.scrapeCanvasUsers(userEmail)
+  .then((users) => {
+    if (users.length === 0) throw 'No user was found in Canvas for that email.';
```
Comment (jrauschenberg) [inline jobs/scheduling/helpers/updateCanvas.js]: Maybe good to specify the email here.

File: jobs/scheduling/helpers/updateCanvas.js [hunk 3]
```diff
@@ -105,24 +105,97 @@ canvas.updateUserGrade = function(user, course, assignment, grade) {
...
+
+  // request courses and find correct courseId;
+  promiseCourseID = promiseUserID.then((userID) => {
+    return canvas.scrapeCanvasEnrollment(userID);
+    })
+    .then((courses) => {
+    courses = courses.filter((course) => {
+      return course.enrollment_state === 'active';
+    });
+    if (courses.length === 0) throw 'No active courses for that user were found in Canvas.';
```
Comment (jrauschenberg) [inline jobs/scheduling/helpers/updateCanvas.js]: Again, the user ID could be useful here.

File: jobs/scheduling/helpers/updateCanvas.js (hunk 2 above)
Comment (apackin) [inline jobs/scheduling/helpers/updateCanvas.js]: This throw sends us into the catch which adds the message `Error finding user with email ${userEmail} in Canvas`

File: www/canvas/helpers.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  let qWithBolding = '';
+
+  const pairedQtoA = formResponse.definition.fields
+                      .sort((a,b) => a.id-b.id)
+                      .map(formatHTML);
+
+  function formatHTML (question, idx) {
+    qWithBolding = question.title.replace('Texter: ', '<b>Texter: </b>');
```
Comment (jrauschenberg) [inline www/canvas/helpers.js]: If you're only using qWithBolding inside this function, you should define it inside this function.

**Comment by jrauschenberg** [comment]
Minor stuff, otherwise the proof is in the pudding. It works well. Good to go. 👍

File: www/canvas/helpers.js (hunk 4 above)
Comment (apackin) [inline www/canvas/helpers.js]: formatHTML is run inside of the .map for each element so I don't want to re declare it each time

**Comment by tongxiang** [comment]
Good canvas API call refactoring, good error handling by emailing Heather. 

👍

**Pull request merged**